src/clr-angular/forms/datepicker/date-input.ts

```diff
@@ -0,0 +1,195 @@
+    /**
+     * Subscriptions to all the services and queries changes
+     */
+    private _subscriptions: Subscription[] = [];
+
+    constructor(@Optional() private container: ClrDateContainer, vcr: ViewContainerRef, private elRef: ElementRef,
+                private renderer: Renderer2, @Self() @Optional() private _ngControl: NgControl,
+                @Optional() private _dateIOService: DateIOService,
+                @Optional() private _dateNavigationService: DateNavigationService,
+                @Optional() private _datepickerEnabledService: DatepickerEnabledService) {
+        super(ClrDateContainer, vcr);
+    }
+
+    /**
+     * 1. Populate services if the date container is not present.
+     * 2. Initialize Subscriptions.
+     * 3. Process User Input.
+     */
+    ngOnInit() {
+        super.ngOnInit();
+        if (!this.container) {
+            this.populateContainerServices();
+        }
+        this.initializeSubscriptions();
+        this.processDate(this.dateValueOnInitialLoad);
+    }
+
+    /**
+     * Write the initial input set by the user on to the input field.
+     */
+    ngAfterViewInit() {
+        // I don't know why I have to do this but after using the new HostWrapping Module I have to delay the processing
+        // of the initial Input set by the user to here.  If I do not 2 issues occur:
+        // 1. the Input setter is called before ngOnInit. ngOnInit initializes the services without which the setter
+        // fails
+        // 2. The Renderer doesn't work before ngAfterViewInit
+        //(It used to before the new HostWrapping Module for some reason).
+        // I need the renderer to set the value property on the input to make sure that if the user has supplied a Date
+        // input object,  we reflect it with the right date on the input field using the IO service.  I am not sure if
+        // these are major issues or not but just noting them down here.
+        if (this._dateIOService.inputDate) {
+            this.writeDateStrToInputField(this._dateIOService.inputDate);
+        }
+        this.initialLoad = false;
+    }
+
+    /**
+     * Unsubscribes from the subscriptions.
+     */
+    ngOnDestroy() {
+        this._subscriptions.forEach((sub: Subscription) => sub.unsubscribe());
+    }
+
+    /**
+     * Populates the services from the container component.
+     */
+    private populateContainerServices(): void {
+        this._dateIOService = this.getProviderFromContainer(DateIOService);
+        this._dateNavigationService = this.getProviderFromContainer(DateNavigationService);
+        this._datepickerEnabledService = this.getProviderFromContainer(DatepickerEnabledService);
+    }
+
+    /**
+     * Initialize DateIO Subscriptions
+     */
+    private initializeSubscriptions(): void {
+        if (this._dateIOService) {
+            this._subscriptions.push(this._dateIOService.dateStrUpdated.subscribe((dateStr) => {
```

Comment:
DateNavigation also has current selected values, should consolidate the implementation to have one source of truth.